File: babel-rs/asd/src/kad/bucket.rs
use super::{KadID, Peer};

#[derive(Debug, Clone)]
pub struct Bucket {
    peers: Vec<Peer>,
}

impl Bucket {
    pub fn new(k: u32) -> Self {
        Self {
            peers: Vec::with_capacity(k as usize),
        }
    }

    pub fn insert(&mut self, peer: Peer) -> Option<usize> {
        if self.is_full() || self.contains(&peer.id) {
            return None;
        }
        let index = self.peers.len();
        self.peers.push(peer);
        Some(index)
    }

    pub fn get(&self, index: usize) -> Option<&Peer> {
        self.peers.get(index)
    }

    pub fn remove(&mut self, index: usize) -> Option<Peer> {
        if index >= self.peers.len() {
            return None;
        }
        Some(self.peers.swap_remove(index))
    }

    pub fn remove_by_id(&mut self, id: &KadID) -> Option<Peer> {
        let index = self.find_peer_index(id)?;
        Some(self.peers.swap_remove(index))
    }

    pub fn len(&self) -> usize {
        self.peers.len()
    }

    pub fn is_empty(&self) -> bool {
        self.peers.is_empty()
    }

    pub fn is_full(&self) -> bool {
        self.peers.len() == self.peers.capacity()
    }

    pub fn contains(&self, id: &KadID) -> bool {
        self.find_peer_index(id).is_some()
    }

    pub fn iter(&self) -> impl Iterator<Item = &Peer> {
        self.peers.iter()
    }

    pub fn extend(&mut self, peers: impl IntoIterator<Item = Peer>) {
        for peer in peers {
            if self.insert(peer).is_none() {
                break;
            }
        }
    }

    pub fn drain_filter<'a>(
        &'a mut self,
        mut f: impl FnMut(&Peer) -> bool + 'a,
    ) -> impl Iterator<Item = Peer> + 'a {
        self.peers.drain_filter(move |p| f(p))
    }

    fn find_peer_index(&self, id: &KadID) -> Option<usize> {
        self.peers.iter().position(|p| p.id == *id)
    }
}

impl std::ops::Index<usize> for Bucket {
    type Output = Peer;

    fn index(&self, index: usize) -> &Self::Output {
        self.get(index).unwrap()
    }
}

File: src/main/java/asd/protocols/overlay/kad/routing/Bucket.java
package asd.protocols.overlay.kad.routing;

import asd.protocols.overlay.kad.KadID;
import asd.protocols.overlay.kad.KadPeer;

import java.util.Arrays;
import java.util.Iterator;
import java.util.stream.Stream;

public class Bucket implements Iterable<KadPeer> {
	private final KadPeer[] peers;
	private int size;

	public Bucket(int k) {
		this.peers = new KadPeer[k];
		this.size = 0;
	}

	public boolean add(KadPeer peer) {
		if (this.isFull()) {
			return false;
		}
		var index = this.findPeerIndex(peer.id);
		if (index != -1) {
			return false;
		}
		this.peers[this.size] = peer;
		this.size += 1;
		return true;
	}

	public boolean contains(KadID id) {
		return this.findPeerIndex(id) != -1;
	}

	/**
	 * Perform a swap-remove on `index`.
	 *
	 * @param index Index of the peer to remove
	 * @return The peer at `index`
	 */
	public KadPeer remove(int index) {
		if (index >= this.size) {
			throw new IndexOutOfBoundsException();
		}
		var peer = this.peers[index];
		this.peers[index] = this.peers[this.size - 1];
		this.peers[this.size - 1] = null;
		this.size -= 1;
		return peer;
	}

	public boolean removeByID(KadID id) {
		var index = this.findPeerIndex(id);
		if (index == -1) {
			return false;
		}
		this.remove(index);
		return true;
	}

	public KadPeer get(int index) {
		if (index >= this.size) {
			throw new IndexOutOfBoundsException();
		}
		return this.peers[index];
	}

	public int size() {
		return this.size;
	}

	public boolean isEmpty() {
		return this.size == 0;
	}

	public boolean isFull() {
		return this.size == this.peers.length;
	}

	public Stream<KadPeer> stream() {
		return Arrays.stream(this.peers, 0, this.size);
	}

	private int findPeerIndex(KadID id) {
		for (int i = 0; i < this.size; ++i)
			if (this.peers[i].id.equals(id)) {
				return i;
			}
		return -1;
	}

	@Override
	public Iterator<KadPeer> iterator() {
		return new Iterator<KadPeer>() {
			private int index = 0;

			@Override
			public boolean hasNext() {
				return this.index < Bucket.this.size;
			}

			@Override
			public KadPeer next() {
				var peer = Bucket.this.peers[this.index];
				this.index += 1;
				return peer;
			}
		};
	}
}
